board.py

```diff
@@ -2,79 +2,215 @@
+            self._set_entity_at_cell(snake.head, EntityId.HEAD)
+
+    def is_snake_id_at_cell(self, snake_id, cell):
+        """Return True if Entity at Cell is a Snake and has snake_id."""
+        for snake in self.get_snakes():
+            if snake.id == snake_id:
+                return cell in snake.head + snake.body
+        return False
+
+    def get_entity_at_cell(self, cell):
+        """Get EntityId at Cell."""
+        return self.get_entity_at_pos(cell.x, cell.y)
+
+    def get_entity_at_pos(self, x, y):
+        """Get EntityId at position."""
+        return self.board[y - 1][x - 1]
+
+    def get_object_at_cell(self, cell):
+        """Get object at Cell."""
+        return self.get_object_at_pos(cell.x, cell.y)
+
+    def get_object_at_pos(self, x, y):
+        """Get object at position."""
+        entity_at_pos = self.get_entity_at_pos(x, y)
+        if entity_at_pos is EntityId.EMPTY:
+            return None
+        elif entity_at_pos in [EntityId.HEAD, EntityId.BODY]:
+            return None
+        elif entity_at_pos is EntityId.FOOD:
+            return None
+
+    def _set_entity_at_cell(self, cell, entity_id):
+        """Set EntityId at Cell."""
+        self._set_entity_at_pos(cell.x, cell.y, entity_id)
+
+    def _set_entity_at_pos(self, x, y, entity_id):
+        """Set EntityId at position."""
+        self.board[y - 1][x - 1] = entity_id
+
+    def cell_within_bounds(self, cell):
+        """Return True if Cell is within bounds of the game grid."""
+        return self.pos_within_bounds(cell.x, cell.y)
+
+    def pos_within_bounds(self, x, y):
+        """Return True if position is within bounds of the game grid."""
+        return 0 <= x <= self.width - 1 and 0 <= y <= self.height - 1
+
+    def update_board(self, snake_id_move_mapping):
+        """State transition for board. Applies moves to all live snakes.
+
+        The Grid is invalidated until the end of this method.
+        Order of events:
+        1. Move snakes
+        2. Snakes dying
+        3. Update world
+        4. Update grid
+        """
+
+        self._move_snakes(snake_id_move_mapping)
+        self._resolve_deaths()
+        self._update_board_after_move()
+        self._build_grid()
+        return self
+
+    def _move_snakes(self, snake_id_move_mapping):
+        """Apply moves to snakes."""
+        for snake_id, move in snake_id_move_mapping.items():
+            self.snakes[snake_id].apply_move(move)
+
+    def _resolve_deaths(self):
+        """Resolve deaths from starvation, wall collision, and snake collision.
+
+        Snakes that die the current turn are moved from snakes to dead_snakes.
+        """
+        starved, not_starved = set(), set()
+        collided_with_wall, not_collided_with_wall = set(), set()
+        collided_with_snake, not_collided_with_snake = set(), set()
+        for snake in self.get_snakes():
+            (starved, not_starved)[snake.health_points > 0].add(snake)
```

Comment:
is this tuple logic really necessary? `not_starved` is simply `snakes - starved`.